### DISCUSSION TITLE
...

### DISCUSSION THREAD

**Discussion by ghost** [NvChad/NvChad#1295] [Q&A] [answered]

**Comment by siduck** [comment]
@zbirenbaum might guide you

**Comment by zbirenbaum** [comment]
Try completely taking out the config. Give it an empty table. Try using copilot.lua+copilot-cmp after you get it setup. It requires that you have the config file generated by copilot.vim but after that you don't need it installed. I wrote those and I can provide more support that way.

**Reply by Thanatermesis** [reply]
is there any howto? I was able to use copilot.lua but I have no way to configure successfully copilot-cmp

**Reply by siduck** [reply, marked as answer]
hi @Thanatermesis , try this 

```lua

  {
    "hrsh7th/nvim-cmp",
    dependencies = {
      {
        "zbirenbaum/copilot-cmp",
        config = function()
          require("copilot_cmp").setup()
        end,
      },
    },
    opts = {
      sources = {
        { name = "nvim_lsp" },
        { name = "luasnip" },
        { name = "buffer" },
        { name = "nvim_lua" },
        { name = "path" },
        { name = "copilot" },
      },
    },
  },
  ```

**Reply by Thanatermesis** [reply]
Where exactly I need to put this code? I tried in different places but Im unable to make it working

**Reply by siduck** [reply]
@Thanatermesis its a plugin definition, pls put it in your custom plugins table. If you dont know where it is then pls read our docs

**Reply by Thanatermesis** [reply]
Thank you! this worked now

**Reply by luizkowalski** [reply]
Hey, did this work fine for you all? I copied and pasted just as is after trying different configurations and it _almost_ worked. 

now and then, it shows a window like this:

<img width="316" alt="image" src="https://user-images.githubusercontent.com/112706/233808579-84b80d29-fb59-453f-a460-32bfe5c84a4a.png">

and when I try to <kb>tab<kb> to get the copilot completion, it just goes to the snippet and back. 

I wonder if it is possible to disable whatever plugin is creating this panel or get this to work smoother, maybe making the copilot option shown in the panel too?

I assume it is safe to comment out some of the sources

**Reply by Thanatermesis** [reply]
Hey @luizkowalski I was able to make copilot working on the menu (inside it) using the code that was recommended to me before

However, I don't like having it in the menu 🤔 I prefer to have it displaying directly all the time in my code so i can easily pick it if i like it (didn't played much with it but this is my impression for now), but then, there's the problem:

* if you have it displayed like in your screnshot, it becomes unusable because the menu hiddens it, I tried to put the menu in the "top" before my actual line but I don't see this configuration anywhere, so my only option is to remove entirely the autocompletion menu (and trigger it using Tab instead, for example)

still trying to figure out what is the best configuration

**Reply by luizkowalski** [reply]
yeah, that's too bad. for me, I rather have it outside the menu and I got it working by commenting out these configs

```
opts = {
      sources = {
       --  { name = "nvim_lsp" },
       -- { name = "luasnip" },
        -- { name = "buffer" },
        -- { name = "nvim_lua" },
        { name = "path" },
        { name = "copilot" },
      },
```

**Reply by luizkowalski** [reply]
@Thanatermesis 

Scratch that part above, I found what really worked: 

Step 1: I disabled any source in the `nvim-cmp` plugin, I hate that panel

```
  {
    "hrsh7th/nvim-cmp",
    opts = {
      sources = {},
    },
  },
```

Step 2: Since I also removed the copilot.lua dep from the `nvim-cmp`, I added `copilot.vim`

```
  {
    "github/copilot.vim",
    lazy = false,
  },
```

that was it, now copilot shows up exactly where I want it: in the code, not in the panels

I'm not sure why the copilot.lua plugin doesn't work like that. Their readme says that the plugin is a "pure lua replacement for [github/copilot.vim](https://github.com/github/copilot.vim)." but it doesn't behave like it. anyway, pretty happy with it now

**Reply by Thanatermesis** [reply]
Ah that's interesting, in what sense you think its different copilot.vim VS copilot.lua ? I also thought that it looked to be better the .vim version (showing better code suggestions) but I only tried it a bit so I cannot confirm this was true. What is your experience, I mean what you didn't liked from the .lua version?

**Reply by luizkowalski** [reply]
the lua version just...didn't work? suggestions didn't show up depending on the config, the default config is counterintuitive (obv this is just my opinion) and there is much more setup needed than the vim one (I mean, I just added the vim plugin version and off I went)

I tried to setup the lua version to work just like the vim version (showing up suggestions on the left, accept with tab) but couldn't get it to work to save my life

gonna stick with the vim one for now

**Reply by Thanatermesis** [reply]
for me it worked fine, i set it to suggest all the time (so text appeared ghosted), next/prev suggestion shows with Alt + ] or [, and to accept was Alt + l

Try to make it working because I think is better to just use Lua plugins in nvim

**Reply by Thanatermesis** [reply]
@luizkowalski This is how I did it:

- In plugins/configs/cmp.lua, add "autocomplete = false" in "local options -> completions" block
- Same file, mappings, search for the <Tab> one and where it says "fallback" put cmp.complete()

_Important: these entries should be put in the "custom" directory one instead, but I didn't check how to do that at the moment, share me your code if you did_

These options removes the menu automatically and only shows it when you press Tab (there's more fine-tuned options but basically these does that), then add all these 3 "copilot" related blocks in your 'custom/plugins.lua' file: http://sprunge.us/wa3fVi

**Reply by luizkowalski** [reply]
wow thanks for the snippet!

**Reply by Thanatermesis** [reply]
no problem :) keep me updated if works the same for you or there's a difference between (like different quality of results)

**Comment by ghost** [comment]


**Reply by siduck** [reply]
@JediMaster25 i hope you are aware of nvim lua configs work + you know lua. Setting up copilot's just installing and configuring copilot-vim & cmp-copilot plugin

**Comment by ghost** [comment]


**Reply by siduck** [reply]
its chadrc and not chardrc

**Comment by mubashirjamali101** [comment]
### Someone please write a guide instead of these broken-context short hints.

Can someone please tell me where, how and what to change. I tried above everything, nothing seems to work.

Thank you for the help, I am new to nvim.

**Reply by siduck** [reply]
@mubashirjamali101 what are you trying to do?

please read all of nvchad.com or you'll never be able to do it by yourself :/

**Reply by mubashirjamali101** [reply]
Hi! @siduck

Thank you so much for the reply and all the work you do. We appreciate your efforts.

I was bit frustrated with setting up GitHub copilot in the nvim (vanilla NvChad setup).

Can you guide us from the start to finish, how to set it up to get the completions like it does in VS Code and other IDEs.

My current situation is that, I am not even able to run `:Copilot setup` or `:Copilot auth`. It doesn't even load for me.

My system (output of running `$ neofetch`):
```bash
$ neofetch
            .-/+oossssoo+/-.               mj@mjpc 
        `:+ssssssssssssssssss+:`           ------- 
      -+ssssssssssssssssssyyssss+-         OS: Ubuntu 22.04.3 LTS x86_64 
    .ossssssssssssssssssdMMMNysssso.       Host: 82KB Lenovo V15 G2 ITL 
   /ssssssssssshdmmNNmmyNMMMMhssssss/      Kernel: 6.2.0-36-generic 
  +ssssssssshmydMMMMMMMNddddyssssssss+     Uptime: 1 hour, 11 mins 
 /sssssssshNMMMyhhyyyyhmNMMMNhssssssss/    Packages: 2550 (dpkg), 53 (brew), 20 
.ssssssssdMMMNhsssssssssshNMMMdssssssss.   Shell: bash 5.1.16 
+sssshhhyNMMNyssssssssssssyNMMMysssssss+   Resolution: 1920x1080 
ossyNMMMNyMMhsssssssssssssshmmmhssssssso   DE: GNOME 42.9 
ossyNMMMNyMMhsssssssssssssshmmmhssssssso   WM: Mutter 
+sssshhhyNMMNyssssssssssssyNMMMysssssss+   WM Theme: Adwaita 
.ssssssssdMMMNhsssssssssshNMMMdssssssss.   Theme: Yaru-dark [GTK2/3] 
 /sssssssshNMMMyhhyyyyhdNMMMNhssssssss/    Icons: Yaru [GTK2/3] 
  +sssssssssdmydMMMMMMMMddddyssssssss+     Terminal: gnome-terminal 
   /ssssssssssshdmNNNNmyNMMMMhssssss/      CPU: 11th Gen Intel i5-1135G7 (8) @  
    .ossssssssssssssssssdMMMNysssso.       GPU: Intel TigerLake-LP GT2 [Iris Xe 
      -+sssssssssssssssssyyyssss+-         Memory: 7444MiB / 19790MiB 
        `:+ssssssssssssssssss+:`
            .-/+oossssoo+/-.                                       
                                                                  
```

Me being technical:
- I have used vim and nvim before, but, never customized it.
- I don't know lua, but I know python.
- I am full-stack dev, (go, python, javascript, react/next)
- If you ask me to run scripts/commands or write lua. I will be happy to do it.
- I am willing to learn the lua for it.

What have I tried:
- I read nvchad docs, github copilot docs, vi stackexchange discussions, also asked GPT/Bard/Rix/Claude.
- I tried some lua scripting in custom plugins folder.

**Reply by siduck** [reply]
@mubashirjamali101 did you put lazy=false for that plugin? please read our plugin docs . And you should obviously know lua for using nvchad! only basics. So read nvchad.com from start to end , in order. 

Also join our discord so you could get help faster

**Reply by Crabzie** [reply]
i'm having a problem doing this too, i authenticated copilot and i'm getting completion i think
![1](https://github.com/NvChad/NvChad/assets/98683647/ec38a306-a792-4edd-9347-012d75010230)

but when i press TAB, that completion doesn't get written in the file, i have to manually type it, when i press tab it start giving me the snippets and going thru them

![2](https://github.com/NvChad/NvChad/assets/98683647/31bed53c-95e5-4ed9-9a51-784915ae988c)

here's my conf structure & init.lua:
![3](https://github.com/NvChad/NvChad/assets/98683647/41d14918-9d97-4a9f-8654-5d3c2cc9054c)

i can provide LSP conf screenshot or any screenshot.

PS: still new to neovim & NvChad ( first day on it actually )
Goal: i want to have clean & correct setup for copilot that won't cause me problems

**Reply by siduck** [reply]
@cunbex  if you able to  make it work on non nvchad config then i can help

**Reply by Crabzie** [reply]
@siduck  is it doable? Or I'm going into a rabbit hole ?
And can you provide any titles I should look into or a starter point ?

**Reply by siduck** [reply]
@cunbex  try with kickstart.nvim or tinyvim, btw  idk if copilot.vim would work well, i'm not sure if cmp override's its functionality i.e with its ghost_text feature, try copilot.lua

**Comment by micahkepe** [comment]
This is the setup that ultimately worked for me for future reference for anyone struggling to set this up like I was:

```
-- OPTION 1: use copilot as a completion source with ghost text
-- This is the default option and replicates the behavior of the official "github/copilot.vim" plugin
-- Adapted from LazyVim's Copilot plugin setup:
--    https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/plugins/extras/coding/copilot.lua
{
  "zbirenbaum/copilot.lua",
  cmd = "Copilot",
  event = "InsertEnter",
  build = ":Copilot auth",
  opts = {
    suggestion = {
      enabled = true,
      auto_trigger = true,
      debounce = 75,
      keymap = {
        accept = "<C-f>", -- accept the current suggestion in full
        accept_word = "<C-w>", -- accept the current word
        accept_line = "<C-l>", -- accept the current line
        -- Cycle through suggestions 
        next = "<C-n>",
        prev = "<C-p>",
        dismiss = "<C-x>",
      },
    },
    panel = { enabled = false },
    filetypes = {
      markdown = true,
      help = true,
      ["."] = true,
      sh = function()
        if string.match(vim.fs.basename(vim.api.nvim_buf_get_name(0)), "^%.env.*") then
          -- disable for .env files
          return false
        end
        return true
      end,
    },
  },
  config = function(_, opts)
    require("copilot").setup(opts)
  end,
},

{
  "nvim-lualine/lualine.nvim",
  optional = true,
  event = "VeryLazy",
  opts = function(_, opts)
    local function copilot_indicator()
      local clients = vim.lsp.get_clients({ name = "copilot" })
      local copilot_active = #clients > 0
      return copilot_active and "󰚩 " or ""
    end
    table.insert(opts.sections.lualine_x, 2, { copilot_indicator })
  end,
},

{
  "hrsh7th/nvim-cmp",
  dependencies = {
    {
      "zbirenbaum/copilot-cmp",
      dependencies = "copilot.lua",
      config = function()
        require("copilot_cmp").setup()
      end,
    },
  },
  opts = function(_, opts)
    local cmp = require "cmp"
    opts.sources = cmp.config.sources {
      { name = "nvim_lsp", group_index = 2 },
      { name = "luasnip", group_index = 2 },
      { name = "buffer", group_index = 2 },
      { name = "nvim_lua", group_index = 2 },
      { name = "path", group_index = 2 },
    }
  end,
},

-- OPTION 2: use copilot as a source for nvim-cmp (no ghost text)
-- To use copilot with nvim-cmp, uncomment the following block
-- This may be a better option if you prefer accepting suggestions with ENTER,
-- however, Copilot's multiline suggestions are harder to read in the cmp menu.

-- {
--   "zbirenbaum/copilot.lua",
--   cmd = "Copilot",
--   event = "InsertEnter",
--   config = function()
--     require("copilot").setup {
--       suggestion = { enabled = true },
--       panel = { enabled = false },
--       filetypes = {
--         markdown = true, -- enable copilot for markdown files
--         ["."] = true,
--         sh = function()
--           if string.match(vim.fs.basename(vim.api.nvim_buf_get_name(0)), "^%.env.*") then
--             -- disable for .env files
--             return false
--           end
--           return true
--         end,
--       },
--     }
--   end,
-- },
--
-- {
--   "hrsh7th/nvim-cmp",
--   dependencies = {
--     {
--       "zbirenbaum/copilot-cmp",
--       config = function()
--         require("copilot_cmp").setup()
--       end,
--     },
--   },
--   opts = {
--     sources = {
--       { name = "nvim_lsp" },
--       { name = "luasnip" },
--       { name = "buffer" },
--       { name = "nvim_lua" },
--       { name = "path" },
--       { name = "copilot" },
--     },
--   },
-- },
```

I made the accept `<C-f>` to not interfere with the tab mapping.
